Added autoreconnect with automatic handling of event filters.

tryptichon opened the pull request:
See the descriptions of the commits.

tryptichon committed:
Bump version to 2.1.5-SNAPSHOT

tryptichon committed:
Added *.iml to gitignore

tryptichon committed:
Avoid builder cluttering on request loop

Split executeBasicRequest from runBasicRequest. Method runRequest uses
runBasicRequest once and only executeBasicRequest after that first try.
This was the builder gets set up only once and will be reused on all
subsequent tries.

tryptichon committed:
Added autoreconnect with automatic handling of event filters

Debugged autoreconnect after network interruption of websocket.
When handling event-ws, an initial list of filter specifications can be
set. These filters get stored in the DefaultWebSocketClient and will be
set right after the event-ws is connected and will be set again on each
reconnect.

tryptichon committed:
Set formatting as required.

mglusiuk commented on java/hiro-client/src/main/java/co/arago/hiro/client/builder/ClientBuilder.java:
```diff
@@ -72,7 +78,11 @@ public ClientBuilder setTimeout(int timeout) {
 
     public ClientBuilder setApiVersion(String version) {
         this.apiVersion = version;
+        return this;
```
👍

fotto commented:
@tryptichon @mglusiuk  I you consider this a bigger rework of the WS handling maybe it's worth setting version to 2.2.x ?

mglusiuk commented on java/hiro-client/src/main/java/co/arago/hiro/client/rest/DefaultWebSocketClient.java:
```diff
@@ -40,18 +42,145 @@
+
+            process(logListener, JSONValue.toJSONString(m));
+
+            if (!reconnect)
```
this is at least confusing if not wrong reconnect then reconnect()

tryptichon replied:
Yeah. I should rename this - this confused me as well.

mglusiuk commented on java/hiro-client/src/main/java/co/arago/hiro/client/rest/DefaultWebSocketClient.java:
```diff
@@ -61,159 +190,120 @@ public DefaultWebSocketClient(String restApiUrl, String urlParameters, TokenProv
+                } else if (nextTryDelay.getSeconds() < 60) {
+                    nextTryDelay = nextTryDelay.plusSeconds(new Random().nextInt(10) + 1);
+                } else {
+                    nextTryDelay = Duration.ofMinutes(new Random().nextInt(10) + 1);
```
do you think it make sense to wait minutes if can not connect?

tryptichon replied:
What would you suggest? My strategy was to try to reconnect waiting 1,2,3 seconds, then add a random number of seconds until the delay reaches one minute and then between one and ten minutes after that. The intent is to avoid cluttering the network with unnecessary requests when the socket will be gone for a long time.

mglusiuk commented on java/hiro-client/src/main/java/co/arago/hiro/client/rest/DefaultWebSocketClient.java:
```diff
@@ -257,6 +347,62 @@ public synchronized void sendMessage(String message) {
+
+    private String getFilterId(Map filter) {
+        String id = String.valueOf(filter.get("filter-id"));
+        if (id == null || StringUtils.isEmpty(id)) {
```
Strings.IsNullOrEmpty

tryptichon replied:
Strings.isNullOrEmpty is not part of Java 8. I could just use "StringUtils.isEmpty(id)" though, because this handles null values as well. DOH! ;D

tryptichon replied:
Just discovered another error here. "String.valueOf(...)" returns the String "null" on a null value. We do NOT want this, but the real null if filter-id is missing. Changed that now.

mglusiuk commented on java/hiro-client/src/main/java/co/arago/hiro/client/rest/DefaultWebSocketClient.java:
```diff
@@ -257,6 +347,62 @@ public synchronized void sendMessage(String message) {
+        LOG.log(Level.INFO, "Remove filter: " + message);
+
+        sendMessage(message);
+        eventFilterMessages.remove(id);
```
maybe make it concurrent collection to avoid client multithreaded access problems?

tryptichon replied:
Sure can do.

tryptichon commented:
> 
> 
> @tryptichon @mglusiuk I you consider this a bigger rework of the WS handling maybe it's worth setting version to 2.2.x ?

In my opinion it fixes a problem that should not have been there in the first place. On the other hand, the API got changed with the addition of the eventFilter stuff. I will set the version to 2.2.0 then.

tryptichon committed:
Added changes according to pull request

* Changed reconnect to isReconnecting.
* Removed unnecessary id == null.
* Made eventFilterMessages a concurrent map.

tryptichon committed:
Bump version to 2.2.0

tryptichon committed:
Changed String.valueOf() to normal cast.

String.valueOf(...) returns the String "null", but we want to have a real
null value if "filter-id" is missing.
